synapse/federation/federation_base.py

```diff
@@ -58,7 +58,12 @@ def __init__(self, hs: "HomeServer"):
 
     @trace
     async def _check_sigs_and_hash(
-        self, room_version: RoomVersion, pdu: EventBase
+        self,
+        room_version: RoomVersion,
+        pdu: EventBase,
+        record_failure_callback: Optional[
+            Callable[[EventBase, str], Awaitable[None]]
+        ] = None,
```

Comment:
See https://github.com/matrix-org/synapse/pull/13815#discussion_r971432765 for why we're using a callback pattern here.